src/osd/OSD.cc

```diff
@@ -7654,9 +7660,9 @@ void OSD::sched_scrub()
 void OSD::resched_all_scrubs()
 {
   dout(10) << __func__ << ": start" << dendl;
-  OSDService::ScrubJob scrub_job;
-  if (service.first_scrub_stamp(&scrub_job)) {
-    do {
+  vector<OSDService::ScrubJob> scrub_job_vec;
```

Comment:
@maydaycc  how about changing `sched_scrub_pg.lower_bound(next)` to `sched_scrub_pg.upper_bound(next)`, and dropping `++iter` in `next_scrub_stamp()`? something like:
```c++
    std::set<ScrubJob>::const_iterator iter = sched_scrub_pg.upper_bound(next);
    if (iter == sched_scrub_pg.cend())
      return false;
    *out = *iter;
    return true;
```

what do you think?

Reply:
This can't avoid reschedule a scrub job twice。

Reply:
@maydaycc i knew. but the net result is the same, right?

Reply:
@tchaikov Yes, the schedule time is calculated from history info.

Reply:
@tchaikov but if `info.stats.stats_invalid && cct->_conf->osd_scrub_invalid_stats` satified, reschedule will stuck in an infinite loop.

Reply:
ahh, you are right. in that case, I think we can have an alternative fix, like:

```c++
  const vector<spg_t> pgs = [this] {
    vector<spg_t> pgs;
    OSDService::ScrubJob job;
    if (service.first_scrub_stamp(&job)) {
      do {
    pgs.push_back(job.pgid);
      } while (service.next_scrub_stamp(job, &job));
    }
    return pgs;
  }();
  for (auto& pgid : pgs) {
    dout(20) << __func__ << ": examine " << pgid << dendl;
    PGRef pg = _lookup_lock_pg(pgid);
    if (!pg)
      continue;
    if (!pg->m_planned_scrub.must_scrub && !pg->m_planned_scrub.need_auto) {
      dout(15) << __func__ << ": reschedule " << pgid << dendl;
      pg->on_info_history_change();
    }
    pg->unlock();
  }
```
what do you think?

Reply:
@tchaikov LGTM, I have another question about c++，can I replace  `auto& pgid : pgs` with `auto pgid : pgs`, since we need not modify it?

Reply:
the reason why i reference `pgid` by reference is that `spg_t` is a non-trivial object, so for two reasons

1. so there is no need to create a temporary copy for using it
2. copying it causes unnecessary overhead, which could have been avoided by adding a `&`.

but, yeah, this is not a critical path. so i don't feel strong on this.

Reply:
@tchaikov tkx,I will test and recommit it again later.